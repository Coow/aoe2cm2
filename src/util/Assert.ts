import Player from "../constants/Player";
import Action from "../constants/Action";
import Exclusivity from "../constants/Exclusivity";

export const Assert = {
    isString(arg: any): void {
        if (typeof arg !== "string") {
            throw new Error("Expected argument to be string, but was " + typeof arg);
        }
    },
    isOptionalString(arg: any): void {
        if (!(typeof arg === "string" || typeof arg === "undefined")) {
            throw new Error("Expected argument to be string or undefined, but was " + typeof arg);
        }
    },
    isStringArray(arg: any): void {
        let failed = false;
        if (Array.isArray(arg)) {
            arg.forEach(function (item) {
                if (typeof item !== "string") {
                    failed = true;
                }
            })
        } else {
            failed = true;
        }
        if (failed) {
            throw new Error("Expected argument to be string[], but was " + arg);
        }
    },
    isOptionalStringArray(arg: any): void {
        if (arg === undefined) {
            return;
        }
        this.isStringArray(arg);
    },
    isPlayer(arg: any) {
        if (!Object.keys(Player).includes(arg)) {
            throw new Error("Expected argument to be a valid Player value, but was " + arg);
        }
    },
    isPlayerOrUndefined(arg: any) {
        if (arg === undefined) {
            return;
        }
        this.isPlayer(arg);
    },
    isImageUrls(arg: any) {
        const argKeys = Object.keys(arg);
        const expectedKeys = ['animated_left', 'animated_right', 'emblem', 'unit'];
        for (let expectedKey of expectedKeys) {
            if (!argKeys.includes(expectedKey)) {
                throw new Error("Expected argument to be a valid ImageUrls value, but was " + arg);
            }
        }
    },
    isImageUrlsOrUndefined(arg: any) {
        if (arg === undefined) {
            return;
        }
        this.isImageUrls(arg);
    },
    isCategoryLimits(arg: any) {
        const argKeys = Object.keys(arg);
        if (!argKeys.includes('pick') || !argKeys.includes('ban')){
                throw new Error("Expected argument to be a valid CategoryLimits value, but was " + JSON.stringify(arg));
        }
        for (let key of Object.keys(arg.pick)) {
            if (typeof arg.pick[key] !== "number") {
                throw new Error("Expected argument to be a valid CategoryLimits value, but was " + JSON.stringify(arg));
            }
        }
        for (let key of Object.keys(arg.ban)) {
            if (typeof arg.ban[key] !== "number") {
                throw new Error("Expected argument to be a valid CategoryLimits value, but was " + JSON.stringify(arg));
            }
        }
    },
    isCategoryLimitsOrUndefined(arg: any) {
        if (arg === undefined) {
            return;
        }
        this.isCategoryLimits(arg);
    },
    isAction(arg: any) {
        if (!Object.keys(Action).includes(arg)) {
            throw new Error("Expected argument to be a valid Action value, but was " + arg);
        }
    },
    isExclusivity(arg: any) {
        if (!Object.keys(Exclusivity).includes(arg)) {
            throw new Error("Expected argument to be a valid Exclusivity value, but was " + arg);
        }
    },
    isBoolean(arg: any) {
        if (![true, false].includes(arg)) {
            throw new Error("Expected argument to be a valid boolean value, but was " + arg);
        }
    }
};