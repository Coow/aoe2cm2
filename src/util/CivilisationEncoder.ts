import Civilisation from "../models/Civilisation";

export const CivilisationEncoder = {
    encodeCivilisationArray(civilisations: Civilisation[]): string {
        let encoded: number = 0;
        const civilisationIds = civilisations.map(value => value.id);
        for (let i = 0; i < Civilisation.ALL.length; i++) {
            if (civilisationIds.includes(Civilisation.ALL[i].id)) {
                encoded += Math.pow(2, i);
            }
        }
        return CivilisationEncoder.toHexString(encoded);
    },

    decodeCivilisationArray(encoded: string): Civilisation[] {
        if (encoded === '3fffffffefff') {
            const civilisations = [...Civilisation.ALL_ACTIVE];
            civilisations.sort((a, b) => a.name.localeCompare(b.name));
            return civilisations;
        }
        try {
            const encodedNumber = parseInt(encoded, 16);
            const binaryString: string = encodedNumber.toString(2);
            const bits: string[] = binaryString.split('');
            const civilisations = [];
            for (let i = 0; i < bits.length; i++) {
                if (bits[i] === '1') {
                    civilisations.push(Civilisation.ALL[bits.length - 1 - i]);
                }
            }
            civilisations.sort((a, b) => a.name.localeCompare(b.name));
            return civilisations;
        } catch (e) {
            console.log("Could not decode value as hex string: ", encoded);
            return [];
        }
    },

    toHexString(input: number): string {
        return input.toString(16);
    }
};
