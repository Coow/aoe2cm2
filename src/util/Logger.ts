import winston from "winston";

export const logger = (typeof window === 'undefined') ? winston.createLogger({
    level: 'info',
    format: winston.format.combine(
        winston.format.splat(),
        winston.format.timestamp(),
        winston.format.json(),
    ),
    transports: [
        new winston.transports.File({filename: 'error.log', level: 'error'}),
        new winston.transports.File({filename: 'aoe2cm.log', maxsize: 10485760}),
        new winston.transports.Console(),
    ]
}) : {
    info: (...args: any[]) => {
        console.log(...args);
    },
    error: (...args: any[]) => {
        console.log(...args);
    },
    warn: (...args: any[]) => {
        console.log(...args);
    },
    debug: (...args: any[]) => {
        console.log(...args);
    }
} as winston.Logger;
