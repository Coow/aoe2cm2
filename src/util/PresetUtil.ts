import Preset from "../models/Preset";
import {Util} from "./Util";
import fs from "fs";
import {logger} from "./Logger";

export const PresetUtil = {
    newPresetId(): string {
        return Util.newDraftId();
    },

    presetExists(presetId: string): boolean {
        const path = `presets/${presetId}.json`;
        return fs.existsSync(path);
    },

    createPreset(preset: Preset) {
        let presetId = PresetUtil.newPresetId();
        while (PresetUtil.presetExists(presetId)) {
            presetId += Util.randomChar();
        }
        preset.presetId = presetId;
        fs.writeFile(`presets/${presetId}.json`, JSON.stringify(preset), (err) => {
            if (err) throw err;
            logger.info(`Preset saved to presets/${presetId}.json`, {presetId});
        });
        return presetId;
    },

    getIdFromUrl(): string | undefined {
        const match: RegExpMatchArray | null = window.location.pathname.match(/\/preset\/([A-Za-z0-9_]+)\/?.*/);
        if (match !== null) {
            return match[1];
        }
        alert('Could not get preset ID from url');
        return undefined;
    },

};
