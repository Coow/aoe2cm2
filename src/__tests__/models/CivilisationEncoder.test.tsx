import {CivilisationEncoder} from "../../util/CivilisationEncoder";
import Civilisation from "../../models/Civilisation";

it('encode empty yields 0', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray([]);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(0));
    expect(encoded).toEqual('0');
});

it('aztecs only yields 1', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray([Civilisation.AZTECS]);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(1));
    expect(encoded).toEqual('1');
});

it('vikings only yields 2^30', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray([Civilisation.VIKINGS]);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(Math.pow(2, 30)));
    expect(encoded).toEqual('40000000');
});

it('all 37 civs yield 2^37-1', () => {
    const encoded = CivilisationEncoder.encodeCivilisationArray(Civilisation.ALL);
    expect(Civilisation.ALL.length).toEqual(37);
    expect(encoded).toEqual(CivilisationEncoder.toHexString(Math.pow(2, 37) - 1));
    expect(encoded).toEqual('1fffffffff');
});

it('decode 0 yields empty array', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('0');
    expect(decoded).toEqual([]);
});

it('decode 1 yields aztecs', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('1');
    expect(decoded).toEqual([Civilisation.AZTECS]);

});

it('decode 2^30 yields vikings', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('40000000');
    expect(decoded).toEqual([Civilisation.VIKINGS]);

});

it('decode 2^37-1 yields all civs', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('1fffffffff');
    expect(decoded).toEqual([...Civilisation.ALL].sort((a, b) => a.name.localeCompare(b.name)));

});

it('decode invalid yields empty array', () => {
    const decoded = CivilisationEncoder.decodeCivilisationArray('invalid');
    expect(decoded).toEqual([]);
});

it('assert order of civilisations has not changed', () => {
    expect(Civilisation.ALL).toMatchSnapshot();
});
