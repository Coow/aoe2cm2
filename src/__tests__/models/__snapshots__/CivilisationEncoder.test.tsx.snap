// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assert order of civilisations has not changed 1`] = `
Array [
  Civilisation {
    "gameVersion": 2,
    "isRandomlyChosenCiv": false,
    "name": "Aztecs",
  },
  Civilisation {
    "gameVersion": 4,
    "isRandomlyChosenCiv": false,
    "name": "Berbers",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Britons",
  },
  Civilisation {
    "gameVersion": 5,
    "isRandomlyChosenCiv": false,
    "name": "Burmese",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Byzantines",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Celts",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Chinese",
  },
  Civilisation {
    "gameVersion": 4,
    "isRandomlyChosenCiv": false,
    "name": "Ethiopians",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Franks",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Goths",
  },
  Civilisation {
    "gameVersion": 2,
    "isRandomlyChosenCiv": false,
    "name": "Huns",
  },
  Civilisation {
    "gameVersion": 3,
    "isRandomlyChosenCiv": false,
    "name": "Incas",
  },
  Civilisation {
    "gameVersion": 3,
    "isRandomlyChosenCiv": false,
    "name": "Indians",
  },
  Civilisation {
    "gameVersion": 3,
    "isRandomlyChosenCiv": false,
    "name": "Italians",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Japanese",
  },
  Civilisation {
    "gameVersion": 5,
    "isRandomlyChosenCiv": false,
    "name": "Khmer",
  },
  Civilisation {
    "gameVersion": 2,
    "isRandomlyChosenCiv": false,
    "name": "Koreans",
  },
  Civilisation {
    "gameVersion": 3,
    "isRandomlyChosenCiv": false,
    "name": "Magyars",
  },
  Civilisation {
    "gameVersion": 5,
    "isRandomlyChosenCiv": false,
    "name": "Malay",
  },
  Civilisation {
    "gameVersion": 4,
    "isRandomlyChosenCiv": false,
    "name": "Malians",
  },
  Civilisation {
    "gameVersion": 2,
    "isRandomlyChosenCiv": false,
    "name": "Mayans",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Mongols",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Persians",
  },
  Civilisation {
    "gameVersion": 4,
    "isRandomlyChosenCiv": false,
    "name": "Portuguese",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Saracens",
  },
  Civilisation {
    "gameVersion": 3,
    "isRandomlyChosenCiv": false,
    "name": "Slavs",
  },
  Civilisation {
    "gameVersion": 2,
    "isRandomlyChosenCiv": false,
    "name": "Spanish",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Teutons",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Turks",
  },
  Civilisation {
    "gameVersion": 5,
    "isRandomlyChosenCiv": false,
    "name": "Vietnamese",
  },
  Civilisation {
    "gameVersion": 1,
    "isRandomlyChosenCiv": false,
    "name": "Vikings",
  },
  Civilisation {
    "gameVersion": 6,
    "isRandomlyChosenCiv": false,
    "name": "Bulgarians",
  },
  Civilisation {
    "gameVersion": 6,
    "isRandomlyChosenCiv": false,
    "name": "Cumans",
  },
  Civilisation {
    "gameVersion": 6,
    "isRandomlyChosenCiv": false,
    "name": "Lithuanians",
  },
  Civilisation {
    "gameVersion": 6,
    "isRandomlyChosenCiv": false,
    "name": "Tatars",
  },
  Civilisation {
    "gameVersion": 7,
    "isRandomlyChosenCiv": false,
    "name": "Burgundians",
  },
  Civilisation {
    "gameVersion": 7,
    "isRandomlyChosenCiv": false,
    "name": "Sicilians",
  },
]
`;
